File: leetcode/linkedlist/Reverse_Linked_List/Iterative/iterative_solution.go
/**
 * Leetcode 206 | Reverse Linked List
 * Difficulty: Easy
 * Editorial: https://dipanjals-notebook.vercel.app/leetcode/linked-list/reverse-a-linked-list/
 */

package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	var prev *ListNode = nil
	curr := head
	for curr != nil {
		currNext := curr.Next
		curr.Next = prev
		prev = curr
		curr = currNext
	}
	return prev
}

File: leetcode/linkedlist/Reverse_Linked_List/Iterative/iterative_solution.py
# Leetcode: 206 | Reverse Linked List
# Difficulty: Easy
# Editorial: https://dipanjals-notebook.vercel.app/leetcode/linked-list/reverse-a-linked-list/

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev, curr = None, head
        while curr:
            currNext = curr.next
            curr.next = prev
            prev = curr
            curr = currNext
        return prev